sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromProtobufSqlFunctions.scala

```diff
@@ -0,0 +1,275 @@
+        s.toString
+    }
+    val descFilePathValue: Option[Array[Byte]] = descFilePath.eval() match {
+      case s: UTF8String => Some(ProtobufHelper.readDescriptorFileContent(s.toString))
+      case null => None
+    }
+    val optionsValue: Map[String, String] = options.eval() match {
+      case a: ArrayBasedMapData if a.keyArray.array.nonEmpty =>
+        val keys: Array[String] = a.keyArray.array.map(_.toString)
+        val values: Array[String] = a.valueArray.array.map(_.toString)
+        keys.zip(values).toMap
+      case _ => Map.empty
+    }
+    val constructor = try {
+      Utils.classForName(
+        "org.apache.spark.sql.protobuf.ProtobufDataToCatalyst").getConstructors().head
+    } catch {
+      case _: java.lang.ClassNotFoundException =>
+        throw QueryCompilationErrors.protobufNotLoadedSqlFunctionsUnusable(
+          functionName = "FROM_PROTOBUF")
+    }
+    val expr = constructor.newInstance(data, messageNameValue, descFilePathValue, optionsValue)
+    expr.asInstanceOf[Expression]
+  }
+}
+
+/**
+ * Converts a Catalyst binary input value into its corresponding Protobuf format result.
+ * This is a thin wrapper over the [[CatalystDataToProtobuf]] class to create a SQL function.
+ *
+ * @param data
+ *   The Catalyst binary input column.
+ * @param messageName
+ *   The Protobuf message name.
+ * @param descFilePath
+ *   The Protobuf descriptor file path.
+ * @param options
+ *   The options to use when performing the conversion.
+ * @since 4.0.0
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(child, messageName, descFilePath, options) - Converts a Catalyst binary input value into its corresponding
+      Protobuf format result.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(s, 'Person', '/path/to/descriptor.desc', map('emitDefaultValues', 'true')) IS NULL FROM (SELECT NULL AS s);
+       [true]
+  """,
+  group = "misc_funcs",
+  since = "4.0.0"
+)
+// scalastyle:on line.size.limit
+case class ToProtobuf(
+    data: Expression,
+    messageName: Expression,
+    descFilePath: Expression,
+    options: Expression) extends QuaternaryExpression with RuntimeReplaceable {
+  override def first: Expression = data
+  override def second: Expression = messageName
+  override def third: Expression = descFilePath
+  override def fourth: Expression = options
+
+  override def withNewChildrenInternal(
+      newFirst: Expression,
+      newSecond: Expression,
+      newThird: Expression,
+      newFourth: Expression): Expression = {
+    copy(data = newFirst, messageName = newSecond, descFilePath = newThird, options = newFourth)
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val messageNameCheck = messageName.dataType match {
+      case _: StringType if messageName.foldable => None
+      case _ =>
+        Some(TypeCheckResult.TypeCheckFailure(
+          "The second argument of the TO_PROTOBUF SQL function must be a constant string " +
```

Comment:
nit:
"The .. argument of .. SQL function must be a (constant string).."

Those 6 error messages might be able to reuse a template string and/or a centralized mapping for clarity. Just a thought for minor improvement no need to be in this pr

Reply:
Thanks! Let me address it from follow-up PR since we have more similar cases from `from_avro` and `to_avro`, and I plan to remove those duplications from follow-ups